server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java

```diff
@@ -235,25 +266,65 @@ private void connectToNodeOrRetry(
                                     managerRefs.decRef();
                                 }));
 
-                                conn.addCloseListener(ActionListener.running(() -> {
-                                    if (connectingRefCounter.hasReferences() == false) {
-                                        logger.trace("connection manager shut down, closing transport connection to [{}]", node);
-                                    } else if (conn.hasReferences()) {
-                                        logger.info(
-                                            """
-                                                transport connection to [{}] closed by remote; \
-                                                if unexpected, see [{}] for troubleshooting guidance""",
-                                            node.descriptionWithoutAttributes(),
-                                            ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
-                                        );
-                                        // In production code we only close connections via ref-counting, so this message confirms that a
-                                        // 'node-left ... reason: disconnected' event was caused by external factors. Put differently, if a
-                                        // node leaves the cluster with "reason: disconnected" but without this message being logged then
-                                        // that's a bug.
-                                    } else {
-                                        logger.debug("closing unused transport connection to [{}]", node);
+                                conn.addCloseListener(new ActionListener<Void>() {
+                                    @Override
+                                    public void onResponse(Void ignored) {
+                                        final NodeConnectionHistory hist = new NodeConnectionHistory(node.getEphemeralId(), null);
+                                        nodeHistory.put(conn.getNode().getId(), hist);
                                     }
-                                }));
+
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        final NodeConnectionHistory hist = new NodeConnectionHistory(node.getEphemeralId(), e);
+                                        nodeHistory.put(conn.getNode().getId(), hist);
+                                    }
```

Comment:
Do we want to store the connection history even when `conn.hasReferences() == false` ? I'm not 100% familiar with this code, but I wonder if we might get the occasional ungraceful disconnect after we've released all our references?

I guess in that case we would eventually discard the entry via `retainConnectionHistory` anyway.

Do we need to be careful with the timing of calls to `retainConnectionHistory` versus the these close handlers firing? I guess any entries that are added after a purge would not survive subsequent purges.